src/locale.js

```diff
@@ -123,19 +167,45 @@ export default class Locale {
 
 	/**
 	 * Base for the {@link #t} method.
-	 *
 	 * @private
 	 */
-	_t( str, values ) {
-		let translatedString = translate( this.uiLanguage, str );
+	_t( msgid = '', values = [] ) {
+		return translate( this.uiLanguage, msgid )
+			.replace( /%(\d+)/g, ( match, idx ) => values[ idx ] );
+	}
+
+	/**
+	 * Base for the {@link #ct} method.
+	 * @private
+	 */
+	_ct( msgctxt = '', msgid = '', values = [] ) {
+		const msg = translate( this.uiLanguage, `${ msgid } [context: ${ msgctxt }]` );
+		const regex = /%(\d+)/g;
+		const parts = [];
+
+		while ( true ) {
+			const prevIndex = regex.lastIndex;
+			const match = regex.exec( msg );
+			if ( match == null ) {
+				parts.push( msg.substring( prevIndex ) );
+				break;
+			}
+			parts.push( msg.substring( prevIndex, match.index ) );
+			parts.push( values[ match[ 1 ] ] );
+		}
 
-		if ( values ) {
-			translatedString = translatedString.replace( /%(\d+)/g, ( match, index ) => {
-				return ( index < values.length ) ? values[ index ] : match;
-			} );
+		if ( values.some( value => typeof value == 'object' ) ) {
+			return parts;
 		}
+		return parts.join( '' );
+	}
 
-		return translatedString;
+	/**
+	 * Base for the {@link #ctn} method.
+	 * @private
+	 */
+	_ctn( msgctxt = '', msgid = '', msgidPlural = '', value = 0 ) {
+		return this._ct( msgctxt, value == 1 ? msgid : msgidPlural, [ value ] );
```

Comment:
Will it work? What kind of JSON structure do you expect if `_ct` is used like that? How `translate` works? Assume that the language is Polish (translation keys are in English, of course).

Also, we are giving up a possibility to use multiple placeholders, which is not ideal. How about `_ctn( msgctxt = '', msgid = '', msgidPlural = '', quantity = 0, values = null )` and if `values` is not set, then it becomes `[ quantity ]?`

Example usage: `ctn( 'Sent x of y files', 'file', '%0 of %1 files', 4, [ 1, 4 ] );`

Reply:
I think that translate function could take quantity as an additional parameter and return a proper string from the JSON language file.

Reply:
The plural function usually takes only one parameter. I will implement proper plural selection after I change the webpack plugin